### DISCUSSION TITLE
2023.04.27 - Community Meeting

### DISCUSSION THREAD

**Discussion by mrengineer7777** [espressif/arduino-esp32#8062] [General]
Hello to all Arduino ESP32 Core Users 🚀 

**Meeting purpose,**
we appreciate all of your contribution and help and we would like to improve your User experience on this Project. 🤩

As a part of this improvement we would like to uncover our internal Development to you, have clear and transparent planning and also listen to your requests, ideas and valuable Feedback.

To fulfill this, we would like to held Bimonthly Community Meeting where we can share main updates and also have an open discussion with you.

---

**💥Tenth Community Meeting will take place on Thursday 27/04/2023 at 4 pm GMT+1 time zone. (1 hour length) We will use Google Meet.**

---

We want to have some idea about how many of you would be interested in joining these Meetings so instead of sharing directly link to Online Meeting itself, it will work by invitation.

**If you want to be added to upcoming Meeting and receive invitation please send an email to [vojtech.bartoska@espressif.com](mailto:vojtech.bartoska@espressif.com).** 

If you already participated on any Community Meeting, you are automatically invited to upcoming Meeting and it's up to you if you join us again or not.

---

### Agenda will be specified below in the comments:  (feel free to add any topic or idea)

1) Development updates & plans
- 2.0.8 release
-- cover #7941 & #7566 
- release plan for 2.0.X versions based on IDF 4.4.X
- status update of 3.0.0 release
-- branches explanation
- CI updates
-- external Libraries testing framework
-- New boards Test
2) Community Topics
- Communication 
- Automatic code formatting
- Breaking change #7942
- Documentation
-- release notes
-- Official Documentation pages
-- API documentation 
- Clarify licensing 
- Issue labeling and cleanup 
- PR Cleanup
3) Open Discussion

---

See you on the Meeting!

**Comment by VojtechBartoska** [comment]
Hi @mrengineer7777, thanks for opening the thread, have edited the content to our usual description. I'm okay with changing the title. 👍

**Reply by VojtechBartoska** [reply]
Invitation have been sent, feel free to add topics for discussion 💯

**Comment by mrengineer7777** [comment]
Topic: Status of release 2.0.8

**Reply by VojtechBartoska** [reply]
added to agenda

**Comment by mrengineer7777** [comment]
Topic: Moving master to 3.0.0.
Currently PRs are targeting master, 2.x, or 3.0.0.  It's confusing!

Discussion:
A reasonable expectation is that `master` should always compile and work correctly.
2.x branch has already diverged from master
3.x branch may not compile correctly and still needs a lot of library fixes

**Reply by VojtechBartoska** [reply]
Also this will be explained

**Comment by mrengineer7777** [comment]
Topic: Status of release 3.0.0

**Reply by VojtechBartoska** [reply]
added to agenda

**Comment by mrengineer7777** [comment]
Topic: issue labeling and cleanup

Overwhelming number of issues makes it difficult to find actual bugs.
I will suggest new labels and perhaps delete old ones.
Some issues are related to user code implementation and may not be bugs.
Lots of "for reference" issues.  These should be added to documentation and closed out.

**Reply by VojtechBartoska** [reply]
Added to agenda.

Update about "Type: For reference issues", we have started transfering those issues into the Documtation.

Some will be answer directly in related documentation page, like in this https://github.com/espressif/arduino-esp32/pull/8115.

Some will go into "FAQ section".

**Comment by mrengineer7777** [comment]
Topic: PR cleanup

Need help here from Espressif employees!  What would it take to implement or close out old PRs?

**Reply by VojtechBartoska** [reply]
We did last Cleanup in Autumn 2022, we have left around 40 PRs. Some of them serve as a reminders, others need to be processed or investigated again. We will have another Clean-up session in May.

Let's put this at the bottom of our Agenda and if we have time, we can discuss this.

If there is any PR which anyone would like to move forward, don't hesitate to link it here, we can prioritize it.

**Comment by mrengineer7777** [comment]
Topic: Communication

Please talk to us!  We have several active developers in this community that know the ESP32 well enough to improve libraries, like me.  However we don't necessarily know "why" things were done a certain way, so our PRs may not be acceptable for merging.  Need timely feedback on PRs.  Help us to help you.

A live comunications channel like Teams or Slack would be amazing.  Even email would be acceptable if it doesn't get lost.
Gitter `arduino-esp32` channel is used for general user issues.
Gitter `esp32-arduino-devs` (private) channel can be helpful.  Notification of messages is poor and group chat nature means it gets ignored a lot.

**Reply by VojtechBartoska** [reply]
Interesting idea, we are open to new suggestions, adding this to the Agenda.

**Reply by brentru** [reply]
Interested in moving from Gitter to Discord, if applicable.

**Reply by mrengineer7777** [reply]
-1 Vote for Discord.  OK for gaming but not serious work.

**Reply by brentru** [reply]
@mrengineer7777  I understand the sentiment - Discord is what we (adafruit) use for a fairly large community (and CircuitPython development), https://adafru.it/discord.

Do note that moving `esp32-arduino-devs` to Slack (free plan) has a 90-day message retention period.

**Comment by mrengineer7777** [comment]
Topic: Automatic code formatting

As noted in #7975 , an automatic formatter would save time and make our code more readable.

**Reply by VojtechBartoska** [reply]
added into agenda

**Comment by mrengineer7777** [comment]
Topic: Clarify Licensing

As noted in #3121.
Licensing isn't a concern for hobbists.
Is it safe for commercial developers to use code from this repository?
What is the status of adding licensing info to headers?

**Comment by VojtechBartoska** [comment]
Topic: CI updates

- https://github.com/espressif/arduino-esp32/discussions/8041
- https://github.com/espressif/arduino-esp32/pull/8071

**Reply by mrengineer7777** [reply]
Looking forward to hearing more about External Libraries testing

**Comment by mrengineer7777** [comment]
Topic: Breaking Changes
#7942 May break existing projects
#7941 [Merged] Breaks external libraries: ESPASyncWebserver, TASMOTA, etc
#7566 [v2.0.6] PR changed partition structure.  Upgrading firmware in field based on v2.0.5 to v2.0.6/7 causes issues due to changed partitions.  Workaround: use a custom partition definition.

**Reply by tobozo** [reply]
#7942 May break existing projects **and libraries**

I can eventually create a short list of projects and libraries I know will be affected by this, please let me know and I'll edit this comment.

[Edit] Those lists aren't exhaustive.

### Libraries

https://github.com/lovyan03/LovyanGFX (used by [~500 projects](https://tobozo.github.io/LovyanGFX.html))
https://github.com/moononournation/Arduino_GFX (used by [~250 projects](https://tobozo.github.io/Arduino_GFX.html))
https://github.com/m5stack/M5Unified (used by [~150 projects](https://tobozo.github.io/Arduino_GFX.html))
https://github.com/codewitch-honey-crisis/esplay
https://github.com/m5stack/M5GFX
https://github.com/m5stack/M5Stack
https://github.com/m5stack/M5Core2
https://github.com/m5stack/M5StickC
https://github.com/tobozo/ESP32-Chimera-Core
https://github.com/tobozo/ESP32-targz
https://github.com/tobozo/ESP32-USB-Soft-Host
https://github.com/tobozo/M5Stack-SD-Updater

### Projects

https://github.com/G4lile0/ESP32-WiFi-Hash-Monster
https://github.com/lovyan03/M5Stack_LovyanLauncher
https://github.com/tobozo/ESP32-BLECollector
https://github.com/micutil/SetWiFi_Mic
More [random projects in the wild](https://github.com/search?l=C%2B%2B&q=M5Stack_Core_ESP32&type=Code) based on code search.

**Reply by Xylopyrographer** [reply]
Must agree that changing all the partition tables is a significant breaking change and a bit of a puzzle. Adding Insights support should be an option only implemented when intenetionally selected by the developer.

An explanation would be good.

**Reply by mrengineer7777** [reply]
I believe the breaking change in #7941 is the redefinition of `#define F` in WString.h.  I would be tempted to revert the PR.

**Reply by VojtechBartoska** [reply]
Hi all, we'll cover #7941 and #7566 in updates about 2.0.8 release as those PRs have been already merged.

#7942 have been added as a seperate topic and we can shape it a bit as it's planned for 3.0.0 milestone. (there is still room for changes)

**Reply by brentru** [reply]
@mrengineer7777 @VojtechBartoska  Thank you for including #7941 in the agenda for tomorrow!

**Reply by VojtechBartoska** [reply]
Quick update, discussed approach for #7942 is shared in the PR, feel free to left your comments there.

**Comment by Xylopyrographer** [comment]
**Topic: Documentation**

Been using this framework for a few years now. The technical work by the maintainers and contributors is great. Biggest challenge initially and still now is the overall state of documentation.

A few points and suggestions.

**1. Release Notes**

A flat chronological list of all PR's included in a release is of little value.

_Recommendations:_

A. Group and summarize by library, changes in the release. A reference to the PR's pertaining to any given item in the summary could be included.

B. Include as a separate item, a list of all libraries and core changes (i.e. changing all partition files to include the Insights partition???), that have breaking changes. Then in the note for that library state what those breaking changes are. A great template is what is done by the Espressif-IDF team.

**2. Official Documentation Pages**

It needs to be much much much easier to allow people to contribute. Current method requires that every person who considers contributing to:
- learn the Restructured Text (rst) markup language.
- figure out an environment to create rst documents on your OS (macOS, Linux, Windows) for simple rst previews.
- figure out how to get an environment up and running that replicates the look of the official pages.

All the above are barriers to contributing.

_Recommendation:_

Publish a "Getting Started" document that points to:
- a simple guide to learning rst
- recommended environments by platform for writing rst documents and previewing as they would be seen when finally published in the repository.

**3. API Documentation**

As the primary purpose of this repository is to allow Espressif to sell more devices, it must be seen as a priority to provide developers a clear way to use the framework.

Since "Arduino" is in the very name, it is logical that this repository would leverage the official Arduino documentation first. That is, each arduino-esp32 library should operate as its Arduino namesake. This means the API documentation for the arduino-esp32 version library only needs to detail enhancements (and differences necessitated by architectural differences in the SoC used by Arduino and the ESP32 variants) added by the arduino-esp32 library version.

After that, documentation would be added for each esp32-specific library.

Well, in an ideal world that would be the way it works.

_Recommendation:_

Given the challenge as per item 2. above, and the aversion programmers have to document their work, recommendation is use the `@brief` method of documenting the API in the header file of each library.

To start, make it a requirement that all PR's starting from date X must add an `@brief` statement for each function modified by that PR.

We used to say, "If it isn't documented, it doesn't exist." For example, there are a number of PR's that have added good functionality to a library but that addition seldom makes it's way to any document. So, after a period of time, that effort is essentially lost. Putting it into the `@brief` would at least allow the work to be noted and found in the future. If nothing else, it is a place to park the information until (or if) an section in the official docs pages is created.

While examples are great, they must be supported with good API documentation.

**Reply by VojtechBartoska** [reply]
Thanks for topic proposal, have added it into the agenda.

**Reply by VojtechBartoska** [reply]
Example of @brief: https://github.com/espressif/esp-idf/blob/master/components/driver/rmt/include/driver/rmt_tx.h#L66

**Comment by VojtechBartoska** [comment]
Hi, agenda have been updated in the description.

**Comment by VojtechBartoska** [comment]
I just figured out I forgot to share the slides, here we go.
[10_Tenth_Community_meeting-Arduino-April_2023.pdf](https://github.com/espressif/arduino-esp32/files/11905677/10_Tenth_Community_meeting-Arduino-April_2023.pdf)
